File: modules/face-blur/ios/RTMPoseEngine.swift
import CoreGraphics
import Foundation
import onnxruntime_objc

struct RTMPosePoint {
  let x: CGFloat
  let y: CGFloat
  let confidence: Float
}

private struct PixelImage {
  let width: Int
  let height: Int
  let bytes: [UInt8]

  init(_ image: CGImage) throws {
    width = image.width
    height = image.height
    var storage = [UInt8](repeating: 0, count: width * height * 4)
    guard let context = CGContext(
      data: &storage,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue |
        CGBitmapInfo.byteOrder32Big.rawValue
    ) else {
      throw RTMPoseError.message("A pose-estimation image buffer could not be created.")
    }
    context.interpolationQuality = .none
    context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    bytes = storage
  }

  /// Bilinear RGB sample. OpenCV's affine border value is black.
  func sample(x: Float, y: Float, border: UInt8 = 0) -> (Float, Float, Float) {
    let x0 = Int(floor(x))
    let y0 = Int(floor(y))
    let tx = x - Float(x0)
    let ty = y - Float(y0)

    func pixel(_ px: Int, _ py: Int) -> (Float, Float, Float) {
      guard px >= 0, py >= 0, px < width, py < height else {
        let value = Float(border)
        return (value, value, value)
      }
      let offset = (py * width + px) * 4
      return (
        Float(bytes[offset]),
        Float(bytes[offset + 1]),
        Float(bytes[offset + 2])
      )
    }

    let a = pixel(x0, y0)
    let b = pixel(x0 + 1, y0)
    let c = pixel(x0, y0 + 1)
    let d = pixel(x0 + 1, y0 + 1)
    let top = (
      a.0 + (b.0 - a.0) * tx,
      a.1 + (b.1 - a.1) * tx,
      a.2 + (b.2 - a.2) * tx
    )
    let bottom = (
      c.0 + (d.0 - c.0) * tx,
      c.1 + (d.1 - c.1) * tx,
      c.2 + (d.2 - c.2) * tx
    )
    return (
      top.0 + (bottom.0 - top.0) * ty,
      top.1 + (bottom.1 - top.1) * ty,
      top.2 + (bottom.2 - top.2) * ty
    )
  }
}

private struct DetectionBox {
  let left: Float
  let top: Float
  let right: Float
  let bottom: Float

  var area: Float { max(0, right - left) * max(0, bottom - top) }
}

private enum RTMPoseError: Error {
  case message(String)
}

extension RTMPoseError: LocalizedError {
  var errorDescription: String? {
    switch self {
    case .message(let message): return message
    }
  }
}

/**
 * Exact production implementation of local-pose-debug exp-0012:
 * RTMDet-nano person detection followed by RTMPose-t COCO-17.
 */
final class RTMPoseEngine {
  static let detectorThreshold: Float = 0.30
  static let landmarkThreshold: Float = 0.15

  private static let detectorWidth = 320
  private static let detectorHeight = 320
  private static let poseWidth = 192
  private static let poseHeight = 256
  private static let keypointCount = 17

  private let environment: ORTEnv
  private let detector: ORTSession
  private let pose: ORTSession

  init() throws {
    let detectorPath = try Self.modelPath(named: "rtmdet_nano_person_320")
    let posePath = try Self.modelPath(named: "rtmpose_t_coco17_256x192")
    environment = try ORTEnv(loggingLevel: .warning)
    detector = try ORTSession(env: environment, modelPath: detectorPath, sessionOptions: nil)
    pose = try ORTSession(env: environment, modelPath: posePath, sessionOptions: nil)
  }

  func analyze(_ image: CGImage) throws -> [RTMPosePoint]? {
    let pixels = try PixelImage(image)
    guard let box = try detectLargestPerson(in: pixels) else { return nil }
    return try estimatePose(in: pixels, box: box)
  }

  private func detectLargestPerson(in image: PixelImage) throws -> DetectionBox? {
    let ratio = min(
      Float(Self.detectorHeight) / Float(image.height),
      Float(Self.detectorWidth) / Float(image.width)
    )
    let resizedWidth = Int(Float(image.width) * ratio)
    let resizedHeight = Int(Float(image.height) * ratio)
    var input = [Float](
      repeating: 0,
      count: 3 * Self.detectorWidth * Self.detectorHeight
    )
    let plane = Self.detectorWidth * Self.detectorHeight
    let mean: [Float] = [103.53, 116.28, 123.675]
    let std: [Float] = [57.375, 57.12, 58.395]

    for y in 0..<Self.detectorHeight {
      for x in 0..<Self.detectorWidth {
        let rgb: (Float, Float, Float)
        if x < resizedWidth, y < resizedHeight {
          let sourceX = (Float(x) + 0.5) / ratio - 0.5
          let sourceY = (Float(y) + 0.5) / ratio - 0.5
          rgb = image.sample(x: sourceX, y: sourceY, border: 114)
        } else {
          rgb = (114, 114, 114)
        }
        let index = y * Self.detectorWidth + x
        let bgr = [rgb.2, rgb.1, rgb.0]
        for channel in 0..<3 {
          input[channel * plane + index] = (bgr[channel] - mean[channel]) / std[channel]
        }
      }
    }

    let value = try Self.tensor(input, shape: [1, 3, 320, 320])
    let outputs = try detector.run(
      withInputs: ["input": value],
      outputNames: ["dets"],
      runOptions: nil
    )
    guard let dets = outputs["dets"] else {
      throw RTMPoseError.message("RTMDet did not return its detection tensor.")
    }
    let shape = try dets.tensorTypeAndShapeInfo().shape.map(\.intValue)
    guard shape.count == 3, shape[2] == 5 else {
      throw RTMPoseError.message("RTMDet returned an unexpected detection shape: \(shape).")
    }
    let values: [Float] = try Self.array(from: dets)
    var largest: DetectionBox?
    for index in 0..<shape[1] {
      let offset = index * 5
      guard offset + 4 < values.count, values[offset + 4] > Self.detectorThreshold else {
        continue
      }
      let box = DetectionBox(
        left: max(0, min(Float(image.width - 1), values[offset] / ratio)),
        top: max(0, min(Float(image.height - 1), values[offset + 1] / ratio)),
        right: max(0, min(Float(image.width - 1), values[offset + 2] / ratio)),
        bottom: max(0, min(Float(image.height - 1), values[offset + 3] / ratio))
      )
      if box.area > (largest?.area ?? 0) { largest = box }
    }
    return largest
  }

  private func estimatePose(in image: PixelImage, box: DetectionBox) throws -> [RTMPosePoint] {
    let centerX = (box.left + box.right) * 0.5
    let centerY = (box.top + box.bottom) * 0.5
    var scaleWidth = (box.right - box.left) * 1.25
    var scaleHeight = (box.bottom - box.top) * 1.25
    let aspect = Float(Self.poseWidth) / Float(Self.poseHeight)
    if scaleWidth > scaleHeight * aspect {
      scaleHeight = scaleWidth / aspect
    } else {
      scaleWidth = scaleHeight * aspect
    }

    var input = [Float](repeating: 0, count: 3 * Self.poseWidth * Self.poseHeight)
    let plane = Self.poseWidth * Self.poseHeight
    let mean: [Float] = [123.675, 116.28, 103.53]
    let std: [Float] = [58.395, 57.12, 57.375]
    for y in 0..<Self.poseHeight {
      for x in 0..<Self.poseWidth {
        let sourceX = centerX + (Float(x) - Float(Self.poseWidth) * 0.5) *
          scaleWidth / Float(Self.poseWidth)
        let sourceY = centerY + (Float(y) - Float(Self.poseHeight) * 0.5) *
          scaleHeight / Float(Self.poseHeight)
        let rgb = image.sample(x: sourceX, y: sourceY)
        let index = y * Self.poseWidth + x
        let bgr = [rgb.2, rgb.1, rgb.0]
        for channel in 0..<3 {
          input[channel * plane + index] = (bgr[channel] - mean[channel]) / std[channel]
        }
      }
    }

    let value = try Self.tensor(input, shape: [1, 3, 256, 192])
    let outputs = try pose.run(
      withInputs: ["input": value],
      outputNames: ["simcc_x", "simcc_y"],
      runOptions: nil
    )
    guard let xValue = outputs["simcc_x"], let yValue = outputs["simcc_y"] else {
      throw RTMPoseError.message("RTMPose did not return both SimCC tensors.")
    }
    let xShape = try xValue.tensorTypeAndShapeInfo().shape.map(\.intValue)
    let yShape = try yValue.tensorTypeAndShapeInfo().shape.map(\.intValue)
    guard xShape.count == 3, yShape.count == 3,
          xShape[1] == Self.keypointCount, yShape[1] == Self.keypointCount else {
      throw RTMPoseError.message("RTMPose returned unexpected SimCC tensor shapes.")
    }
    let xValues: [Float] = try Self.array(from: xValue)
    let yValues: [Float] = try Self.array(from: yValue)
    let xBins = xShape[2]
    let yBins = yShape[2]

    return (0..<Self.keypointCount).map { keypoint in
      let xRange = (keypoint * xBins)..<((keypoint + 1) * xBins)
      let yRange = (keypoint * yBins)..<((keypoint + 1) * yBins)
      let xMaximum = xRange.max { xValues[$0] < xValues[$1] } ?? xRange.lowerBound
      let yMaximum = yRange.max { yValues[$0] < yValues[$1] } ?? yRange.lowerBound
      let modelX = Float(xMaximum - xRange.lowerBound) / 2
      let modelY = Float(yMaximum - yRange.lowerBound) / 2
      let originalX = modelX / Float(Self.poseWidth) * scaleWidth + centerX - scaleWidth / 2
      let originalY = modelY / Float(Self.poseHeight) * scaleHeight + centerY - scaleHeight / 2
      return RTMPosePoint(
        x: CGFloat(originalX / Float(max(1, image.width))),
        y: CGFloat(originalY / Float(max(1, image.height))),
        confidence: (xValues[xMaximum] + yValues[yMaximum]) * 0.5
      )
    }
  }

  private static func tensor(_ values: [Float], shape: [NSNumber]) throws -> ORTValue {
    let data = values.withUnsafeBufferPointer { Data(buffer: $0) }
    return try ORTValue(
      tensorData: NSMutableData(data: data),
      elementType: .float,
      shape: shape
    )
  }

  private static func array(from value: ORTValue) throws -> [Float] {
    let data = try value.tensorData() as Data
    guard data.count % MemoryLayout<Float>.stride == 0 else {
      throw RTMPoseError.message("An ONNX output tensor had an invalid byte count.")
    }
    return data.withUnsafeBytes { Array($0.bindMemory(to: Float.self)) }
  }

  private static func modelPath(named name: String) throws -> String {
    let bundles = [Bundle.main, Bundle(for: RTMPoseEngine.self)] + Bundle.allFrameworks
    for bundle in bundles {
      if let path = bundle.path(forResource: name, ofType: "onnx") { return path }
    }
    throw RTMPoseError.message("The bundled \(name).onnx model could not be found.")
  }
}

File: modules/face-blur/android/src/main/java/expo/modules/faceblur/RTMPoseEngine.kt
package expo.modules.faceblur

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.content.Context
import android.graphics.Bitmap
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.security.MessageDigest
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

internal data class RTMPosePoint(
  val x: Float,
  val y: Float,
  val confidence: Float,
)

private data class DetectionBox(
  val left: Float,
  val top: Float,
  val right: Float,
  val bottom: Float,
) {
  val area: Float get() = max(0f, right - left) * max(0f, bottom - top)
}

private class RTMPoseDiagnosticWriter(private val directory: File) {
  init {
    check(directory.mkdirs() || directory.isDirectory) {
      "Could not create the pose diagnostic directory."
    }
  }

  fun writeBytes(values: ByteArray, name: String) {
    File(directory, name).writeBytes(values)
  }

  fun writeFloats(values: FloatArray, name: String) {
    val bytes = ByteBuffer.allocate(values.size * Float.SIZE_BYTES)
      .order(ByteOrder.LITTLE_ENDIAN)
    values.forEach(bytes::putFloat)
    File(directory, name).writeBytes(bytes.array())
  }

  fun writeJson(value: JSONObject, name: String) {
    File(directory, name).writeText(value.toString(2) + "\n")
  }
}

/** Exact Android implementation of local-pose-debug exp-0012. */
internal class RTMPoseEngine(context: Context) : AutoCloseable {
  companion object {
    const val DETECTOR_THRESHOLD = 0.30f
    const val LANDMARK_THRESHOLD = 0.15f

    private const val DETECTOR_MODEL = "rtmdet_nano_person_320.onnx"
    private const val DETECTOR_SHA =
      "8297e829ccc5590c8e2d32d5a211f322a0585fb7467eec85eb12c9525b0b95d6"
    private const val POSE_MODEL = "rtmpose_t_coco17_256x192.onnx"
    private const val POSE_SHA =
      "a6c2f6a3896a4d51131d14d7a80a3d08b50f559af5a58a45d5b098aef510a70f"

    private const val DETECTOR_WIDTH = 320
    private const val DETECTOR_HEIGHT = 320
    private const val POSE_WIDTH = 192
    private const val POSE_HEIGHT = 256
    private const val KEYPOINT_COUNT = 17
  }

  private val environment = OrtEnvironment.getEnvironment()
  private val detector = createSession(
    materializeModel(context, DETECTOR_MODEL, DETECTOR_SHA),
  )
  private val pose = createSession(
    materializeModel(context, POSE_MODEL, POSE_SHA),
  )

  fun analyze(bitmap: Bitmap, diagnosticsDirectory: File? = null): List<RTMPosePoint>? {
    val diagnostics = diagnosticsDirectory?.let(::RTMPoseDiagnosticWriter)
    val pixels = IntArray(bitmap.width * bitmap.height)
    bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
    diagnostics?.writeBytes(rgbaBytes(pixels), "display-rgba.bin")
    val box = detectLargestPerson(pixels, bitmap.width, bitmap.height, diagnostics)
    if (box == null) {
      diagnostics?.writeJson(
        JSONObject()
          .put("imageWidth", bitmap.width)
          .put("imageHeight", bitmap.height)
          .put("detected", false),
        "result.json",
      )
      return null
    }
    val points = estimatePose(pixels, bitmap.width, bitmap.height, box, diagnostics)
    diagnostics?.writeJson(
      JSONObject()
        .put("imageWidth", bitmap.width)
        .put("imageHeight", bitmap.height)
        .put("detected", true)
        .put("detectorBox", JSONArray(listOf(box.left, box.top, box.right, box.bottom)))
        .put("keypoints", JSONArray(points.map {
          JSONArray(listOf(it.x, it.y, it.confidence))
        })),
      "result.json",
    )
    return points
  }

  private fun detectLargestPerson(
    pixels: IntArray,
    width: Int,
    height: Int,
    diagnostics: RTMPoseDiagnosticWriter?,
  ): DetectionBox? {
    val ratio = min(DETECTOR_HEIGHT.toFloat() / height, DETECTOR_WIDTH.toFloat() / width)
    val resizedWidth = (width * ratio).toInt()
    val resizedHeight = (height * ratio).toInt()
    val plane = DETECTOR_WIDTH * DETECTOR_HEIGHT
    val input = FloatArray(3 * plane)
    val sample = FloatArray(3)
    val means = floatArrayOf(103.53f, 116.28f, 123.675f)
    val stds = floatArrayOf(57.375f, 57.12f, 58.395f)
    for (y in 0 until DETECTOR_HEIGHT) {
      for (x in 0 until DETECTOR_WIDTH) {
        if (x < resizedWidth && y < resizedHeight) {
          val sourceX = (x + 0.5f) / ratio - 0.5f
          val sourceY = (y + 0.5f) / ratio - 0.5f
          sampleRgb(pixels, width, height, sourceX, sourceY, 114, sample)
        } else {
          sample.fill(114f)
        }
        val index = y * DETECTOR_WIDTH + x
        input[index] = (sample[2] - means[0]) / stds[0]
        input[plane + index] = (sample[1] - means[1]) / stds[1]
        input[2 * plane + index] = (sample[0] - means[2]) / stds[2]
      }
    }

    diagnostics?.writeFloats(input, "detector-input-f32.bin")

    OnnxTensor.createTensor(
      environment,
      FloatBuffer.wrap(input),
      longArrayOf(1, 3, DETECTOR_HEIGHT.toLong(), DETECTOR_WIDTH.toLong()),
    ).use { tensor ->
      detector.run(mapOf("input" to tensor), setOf("dets")).use { result ->
        @Suppress("UNCHECKED_CAST")
        val batches = result.get("dets").orElseThrow {
          IllegalStateException("RTMDet did not return its detection tensor.")
        }.value as Array<Array<FloatArray>>
        val detections = batches.firstOrNull().orEmpty()
        diagnostics?.writeFloats(
          detections.flatMap { it.asIterable() }.toFloatArray(),
          "detector-output-f32.bin",
        )
        diagnostics?.writeJson(
          JSONObject().put("shape", JSONArray(listOf(1, detections.size, 5))),
          "detector-output-shape.json",
        )
        var largest: DetectionBox? = null
        for (detection in detections) {
          if (detection.size < 5 || detection[4] <= DETECTOR_THRESHOLD) continue
          val box = DetectionBox(
            left = (detection[0] / ratio).coerceIn(0f, (width - 1).toFloat()),
            top = (detection[1] / ratio).coerceIn(0f, (height - 1).toFloat()),
            right = (detection[2] / ratio).coerceIn(0f, (width - 1).toFloat()),
            bottom = (detection[3] / ratio).coerceIn(0f, (height - 1).toFloat()),
          )
          if (box.area > (largest?.area ?: 0f)) largest = box
        }
        return largest
      }
    }
  }

  private fun estimatePose(
    pixels: IntArray,
    width: Int,
    height: Int,
    box: DetectionBox,
    diagnostics: RTMPoseDiagnosticWriter?,
  ): List<RTMPosePoint> {
    val centerX = (box.left + box.right) * 0.5f
    val centerY = (box.top + box.bottom) * 0.5f
    var scaleWidth = (box.right - box.left) * 1.25f
    var scaleHeight = (box.bottom - box.top) * 1.25f
    val aspect = POSE_WIDTH.toFloat() / POSE_HEIGHT
    if (scaleWidth > scaleHeight * aspect) {
      scaleHeight = scaleWidth / aspect
    } else {
      scaleWidth = scaleHeight * aspect
    }
    val plane = POSE_WIDTH * POSE_HEIGHT
    val input = FloatArray(3 * plane)
    val sample = FloatArray(3)
    val means = floatArrayOf(123.675f, 116.28f, 103.53f)
    val stds = floatArrayOf(58.395f, 57.12f, 57.375f)
    for (y in 0 until POSE_HEIGHT) {
      for (x in 0 until POSE_WIDTH) {
        val sourceX = centerX + (x - POSE_WIDTH * 0.5f) * scaleWidth / POSE_WIDTH
        val sourceY = centerY + (y - POSE_HEIGHT * 0.5f) * scaleHeight / POSE_HEIGHT
        sampleRgb(pixels, width, height, sourceX, sourceY, 0, sample)
        val index = y * POSE_WIDTH + x
        input[index] = (sample[2] - means[0]) / stds[0]
        input[plane + index] = (sample[1] - means[1]) / stds[1]
        input[2 * plane + index] = (sample[0] - means[2]) / stds[2]
      }
    }

    diagnostics?.writeFloats(input, "pose-input-f32.bin")

    OnnxTensor.createTensor(
      environment,
      FloatBuffer.wrap(input),
      longArrayOf(1, 3, POSE_HEIGHT.toLong(), POSE_WIDTH.toLong()),
    ).use { tensor ->
      pose.run(mapOf("input" to tensor), setOf("simcc_x", "simcc_y")).use { result ->
        @Suppress("UNCHECKED_CAST")
        val simccX = result.get("simcc_x").orElseThrow {
          IllegalStateException("RTMPose did not return simcc_x.")
        }.value as Array<Array<FloatArray>>
        @Suppress("UNCHECKED_CAST")
        val simccY = result.get("simcc_y").orElseThrow {
          IllegalStateException("RTMPose did not return simcc_y.")
        }.value as Array<Array<FloatArray>>
        val xValues = simccX.firstOrNull()
          ?: throw IllegalStateException("RTMPose returned an empty simcc_x tensor.")
        val yValues = simccY.firstOrNull()
          ?: throw IllegalStateException("RTMPose returned an empty simcc_y tensor.")
        require(xValues.size == KEYPOINT_COUNT && yValues.size == KEYPOINT_COUNT) {
          "RTMPose returned an unexpected keypoint count."
        }
        diagnostics?.writeFloats(
          xValues.flatMap { it.asIterable() }.toFloatArray(),
          "simcc-x-f32.bin",
        )
        diagnostics?.writeFloats(
          yValues.flatMap { it.asIterable() }.toFloatArray(),
          "simcc-y-f32.bin",
        )
        diagnostics?.writeJson(
          JSONObject()
            .put("xShape", JSONArray(listOf(1, xValues.size, xValues.firstOrNull()?.size ?: 0)))
            .put("yShape", JSONArray(listOf(1, yValues.size, yValues.firstOrNull()?.size ?: 0))),
          "pose-output-shapes.json",
        )
        return (0 until KEYPOINT_COUNT).map { keypoint ->
          val xMaximum = xValues[keypoint].indices.maxByOrNull { xValues[keypoint][it] } ?: 0
          val yMaximum = yValues[keypoint].indices.maxByOrNull { yValues[keypoint][it] } ?: 0
          val modelX = xMaximum / 2f
          val modelY = yMaximum / 2f
          RTMPosePoint(
            x = (modelX / POSE_WIDTH * scaleWidth + centerX - scaleWidth / 2f) /
              max(1, width).toFloat(),
            y = (modelY / POSE_HEIGHT * scaleHeight + centerY - scaleHeight / 2f) /
              max(1, height).toFloat(),
            confidence = (xValues[keypoint][xMaximum] + yValues[keypoint][yMaximum]) * 0.5f,
          )
        }
      }
    }
  }


  private fun rgbaBytes(pixels: IntArray): ByteArray {
    val bytes = ByteArray(pixels.size * 4)
    pixels.forEachIndexed { index, pixel ->
      val offset = index * 4
      bytes[offset] = ((pixel shr 16) and 0xff).toByte()
      bytes[offset + 1] = ((pixel shr 8) and 0xff).toByte()
      bytes[offset + 2] = (pixel and 0xff).toByte()
      bytes[offset + 3] = ((pixel ushr 24) and 0xff).toByte()
    }
    return bytes
  }

  override fun close() {
    detector.close()
    pose.close()
  }

  private fun createSession(model: File): OrtSession = OrtSession.SessionOptions().use { options ->
    environment.createSession(model.absolutePath, options)
  }

  private fun sampleRgb(
    pixels: IntArray,
    width: Int,
    height: Int,
    x: Float,
    y: Float,
    border: Int,
    output: FloatArray,
  ) {
    val x0 = floor(x).toInt()
    val y0 = floor(y).toInt()
    val tx = x - x0
    val ty = y - y0

    fun channel(px: Int, py: Int, shift: Int): Float {
      if (px !in 0 until width || py !in 0 until height) return border.toFloat()
      return ((pixels[py * width + px] shr shift) and 0xff).toFloat()
    }

    // Android pixels are ARGB; write RGB into output.
    val shifts = intArrayOf(16, 8, 0)
    for (index in shifts.indices) {
      val shift = shifts[index]
      val top = channel(x0, y0, shift) * (1f - tx) + channel(x0 + 1, y0, shift) * tx
      val bottom = channel(x0, y0 + 1, shift) * (1f - tx) +
        channel(x0 + 1, y0 + 1, shift) * tx
      output[index] = top * (1f - ty) + bottom * ty
    }
  }

  private fun materializeModel(context: Context, assetName: String, expectedSha: String): File {
    val directory = File(context.noBackupFilesDir, "pose-models").apply { mkdirs() }
    val target = File(directory, assetName)
    if (!target.isFile || sha256(target) != expectedSha) {
      val temporary = File(directory, "$assetName.tmp")
      context.assets.open(assetName).use { input ->
        FileOutputStream(temporary).use { output -> input.copyTo(output) }
      }
      check(sha256(temporary) == expectedSha) { "Bundled pose model failed its integrity check: $assetName" }
      if (target.exists()) target.delete()
      check(temporary.renameTo(target)) { "Could not install bundled pose model: $assetName" }
    }
    return target
  }

  private fun sha256(file: File): String {
    val digest = MessageDigest.getInstance("SHA-256")
    file.inputStream().use { input ->
      val buffer = ByteArray(1024 * 1024)
      while (true) {
        val count = input.read(buffer)
        if (count < 0) break
        digest.update(buffer, 0, count)
      }
    }
    return digest.digest().joinToString("") { "%02x".format(it) }
  }
}
